apache/superset
# fix: Reordering echart props to fix confidence interval in Mixed Charts

### fix(echarts): confidence interval does not display correctly for mixed time series charts with negative values

### SUMMARY

The root cause of this problem lies within the order of the arguments sent to the echarts library. When working with time-series data, Echarts enforces specific constraints on how the x-axis bounds are configured. These constraints include:
In other words, the Echarts Library accepts input in this specific format and will malfunction otherwise. Superset does not create the Echarts modal with the options in this order.
 
*  The `xaxis.type` option must be explicitly set to `'time'`.
*  The `upper` and `lower` bound values must be equal to the lower bound.
*  The `upper` bound must be defined before the `lower` bound in the options object.

Currently, the code in Superset that generates the Echarts configuration does not adhere to this required format, leading to the observed malfunction. This pull request rectifies this issue by reordering the `upper` and `lower` bound arguments before they are passed to the Echarts library.

[What is currently being sent](https://echarts.apache.org/examples/en/editor.html?version=5.5.0&code=PYBwLglsB2AEC8sDeAoW7YCICuBnApgKoAqAwpgFyxgBO2-ANGhpgOY0QAmlyzGLAYxhgAhhGgAZEQCN8AGx616Tfv0xz8AMzA8AzAFYVq9Jg6sAFjqoAmAAxHjmMKB4AWe31WZpwMM4C2PHaeAL4OWAAeAIIRELg8qMYmYACeIPg8mABuInL0mOFq0CL--ADiIiA8AIyGnkUl-BLAAiKQMJn-XJwaBfUsIrG4UrIKVEhh_SZd0MA0xBACANYJk0lYMwCS0GD4NDljsB7rmDOZnG0iALLifSf-g-eXV4-h4ZgpMXE8ANpTvOs1J8hiN5AAlXxtKDQHjHQHJNIZKjZXL5QpJU7iOYLZardGOGZzADKIDkEDAEnESOQa3hWEGcVBhwm-K8xVKFSqVFqrIapWarXaMORXU4PQyvIGZNY0BxS3iVE0uQI_1pxkSdI-X2GMnBkKFsMlCPSmQOaP-akJ80WK3GasBmFwpPJlOg1I1dJYuHMwAA7jwlXIVZ77fcsTQSWSKVS8RaBiDdczQxjchAZXKFbBA8GkiFPABdd7OYBySBcgEpkDpaCcYjAABCwE4KUUdEY_0dPv9VCU7ZOtDTrD2mQZ8TennU-CHNYSHbmEHwO0yPo4AC9hLk7hjvX7W8oO6kTcjcAIaCWFKzHfJ8AJnDRfnGTLkL4-sOIsnsCJh_oWOwQNLecxMrOnqYJowgAGIlBAcgtsiADk2y7DQ8EMLAAAS8gfpArRoVEHCbkaWDgTsRIQKu1LVNYRGYEIchzJkADEABsrFsVu8LeHMnB7KQJYMciLFsax37rMmyQuFQcKOGYliGh2FyiA-oERBxDq-OYw4_uOagEBw-CZn86weg6XCZBEAD6FkpOYbQWdgVbDjR7LUpgqk0YpIjKZ6GBGT5xjWHU_kBbYr78L-wWwH5kXoNYADsRHrMEkURcF0UxdYAAciVJMlwWpf56WRdYACcOUhWFGAFT5RXBbo0kZaFKXlVFlXGLo1QtfweX-dVnq1f5ujUW1qg9T5fV0gNPm6K4XUYGNnoTfCU2egYc2xU1-UtStdLuOtsALXS-avktXgAI70DQKTbDxqlSTRwJxDd-B3UcNHkvg_hEqkvTjCNtH8fegnVBBUSZaQJVqaBoAiAI5JwbA1SvuJXiHq5ZJulD27-MAGkBsqfaagOrBDjQrr4AAoh-S49m2NG4KIuLIu5r6OozSzfTQbRTgjmDPljjgY_g30pL9Fb-ZgMNw6ksLIzRIg0PgIgi2LJnQyAsPw7CAB01hy6zn0gLZuDfH9kXqFSKvuiNai-lwYDmJkPhyDx94jXmIb012RIpP4zv40GhOcbgvvO2RFE8MxNFyImIHBZ2u6KgTLWYCAwAmwayLOFUyOqqyasYmZzNWTZdn0b6Tmsy55kC14nneYVNvzUFMXoNUm29dtTexQl3eIx341d63sXZX37cjad6w7fCpX7ePzUjdPgL1XPA-LUPw8davE8b63Q3bwvkVL-sM0H1ti992tY9r0du8xXt18TydNEXXs101i98mgY9uDPa9DUnA-l9H61tzZ0QElgRirh9CuApnFcgKcpZaykvrUCaNMhC1rrpHGeMk6BxosTUm5MqaLisNQOmrMGaw1tJELBXp2ac25qwXm_No6WxAXHCWSCZYoM9gbfwRsRAm0zAXTiQsracJipgO2nAHZOxLK7USwUPZ0hRrpb2ocSwBwIPTTRchw7UijqzGOoxJGgR3N2LMyd_ppwztCTI2clHwhUcYFGoivBF1oc5RoNcPKXAbjVPugU576B3hfYeB1e4RNqGEo-QTR7RNCYfNKQSypjySefOJESV7pNiSk7JnVcnJMbtk4aiS8klM3rNIpmT8mbxbq3GJxTAkRIfuUlKz9Wavyun_L-mof69PuqzIBEizbx3AUDSB-goikGqJlSGiCNbSwRjfVxBDEQYJjPTHBci8E6NZoQvYxDqZkN7PTBhtAmG81wI0E2Mo6EmHERwsZXClnIMRqgzUhtjam3Fj5C2bpRl_PNjI3ZWBnaKPdinEZzzgXx24QjJGkUXGAjUV6DRfstG033OYvRBjI7R1ji8_5FjtFB3-bY8k9is4uFzmJfOwzuDIg0pXUC1dmUO1ZZqeuVBj6-SCQ0mKSLamVNbvFOeFSWnDyyhK5p_VUmypFVKveADgrCs7uEzehTomSvlaUxVGqslVINYPTVe9BWRXVaao1rc2nDytYtTpoFunv1ujUB62pBlvWGbsYBotQHjMBkxCCEE4oQVcK4B5XgEU1E-ZxdBx5BTISjY6HZjs9nkv7GYIhVISE03ITizUVDliMN2MwzINzSh3JhGwwFsL3EnBjdyONDpvlCN-Q2jETz_VmPjqC9N4KFFaWRS2k4FifaYsOGcyheLyKGMJaY4l5iuxkpTpSzOWBHF0tzAWIsuM5A-Feu4pwklvUnFkmQi1LBNBKzANgRWvbHCeQAFq40CEu_pnqP6vWzJmom5JVZ90wOuYA76sAgf8LABWSso0nGkNQzIisGZzHwLACDTifIorEqqHSLAX1vt-MdEIABuIAA)

[What should actually be sent](https://echarts.apache.org/examples/en/editor.html?version=5.5.0&code=PYBwLglsB2AEC8sDeAoWsCuBnApgVQBUBhALljACcMcAaNWAcwogBMzV10BjGMAQwjQAMnwBGOADZlK1Op1gScAMzBkAzAFY5nZgwAWq2ACYADNvRhQZACxn66UcDCWAtmVP0AvtoAeAQR8ILHZ7cgBPEBwyAHJIFxxo81hoPniAcT4QMgBGLVCU-KFgLj5IGBiXVhZFRNC-QKwRcSlkb1DK6GAKAgguAGt2NvkOgEloMBwKADc-FrthwRiWUr4AWUFa4fqlldX66K9tMICgsgBtUI55WGOGpskAJSdSqGgyeeuwCKjYaJmJaiba4dLo9fqDJLoEEUADKIAkEDAQkEPyQQ2u9SC9xaaMhyVSOAyWVguTxBRwRRKZTev0qLGqCTxswgDGgYL6wVgSlmuFC6PQV3ktyxYkez2p7zxX0iMX-gLx0PZENC6Cw8MRyOgqJVnCwemAAHcyNyJLzrrB-ZxoXCEUiUcrzZjGqKcZb0MzWezOSazZxPPQALraSzACSQYmC2CZSLQFgEYAAIWALDC0iotFCesNadkoUoLIYkxiToO6HRikLsZC8i6EBw4xi-uYAC9eLMgar9UbyOmktKftEsFwKKGJB3YLhFFxLBRztFZmOaL9BFNJrhokHM5IcNOutjq9clLwAGKpCASVOwABE0TGEwoiVgAAlJKvICUl35mLMr3ij-MYQgZsfmyIw8R4CQuhiABiAA2eCEPHBwuhYSYiFDKDfjghD4NLP0-ysWAPnQXQDElUJln4OcfEfaInD0SYN0OehcGYHBOQueRI3QVgYh8AB9fiwj0Up-Mgg1GLJAk-KQ2BKL4c4dXQM4jC0EkTE3c1lKMAB2Jdsg0vFtIADn0wylNgFSAE4zM0rSzjUMx1Ls80HOyWyjMstQwOczyHOsDyLIctSDJc64zlsWydTC9AAEdqAoMIxlQnxyPNYUsGSnBUqIvFERwFwYS-RQD3NCDMOiaDrA0awAFEdKIWT0FAPguERS9sh1N1whlX4ES1WSsBcYB6ONHkM3NfMGELChNRwWrVwbHtc3NLB-HBX4aLxNbWr6IqKFKHAGEvecJEXHV-pwIqwhK5ALJatqvneLq8QKkARKwU47q0hQUWu27uOuA1WDAPQYkcCRUIfCz_XNbqswNGEwhcCGxtNCbriwZGIcA4CyFgvEJBdUrMa7NHcE8kBgE-iVfksEA8PkWH8MuHVeM2wThNEjAQEiB8pPiGS8XkxSfpUkLzLF3TAql0zfKCowbPlsXHJl-y1Hc5X1Z80K_LUAKtdczQ1dcyLnOivF4smJLY2ytLrgyrKcuIzh8sK4rtR-8rZyw7Jjz8YyiCsprYAe9qchenV-xiS7BuG0auXGqVdBmuaFvrQwZAx-Qdo26Itp1XO9soQ7juLM7ZMu_7Pa0sOnqIyPHQoHA-GrkmaxAVrw6IgA6IxG-uN6Pq-wH0Crj326BkGwd-CGocZ65maZ7auyRlHQ3J7PdWx0NcZ-AmLuJ76tIRzfKepxFXhiemF7LPkkkB9n89k8khZ1EXLIVkKNBizgVL0kkP8_JGDlrkX-2klZgL1k5KBQUNb6SAXAnWiCVYG1gSrb-4DLJmzARbHUVtEpO3tkKE4mVbbOzyhMd2N0a5lQwj7SqGg_BEGyMZYOnk66XhMAPeQ0c-oojjiNUGZ8o4p0mGnRamdeyF3WsXA6Ewy6_CwAST6rJK5_Qnsfc0nCI5w1ei4d6fBPqclHr9LUbctFaWBiwYRs9Qzzxhp5N2FjTHNU7o9DqMMeG6lXjvFoWdtp-L3vjQmR9TGn0Tujc-NMr501ALfC099WbmifvRSSOpX6_DSfzd-KxRb2VUvpLB_8inANAcUxWpS4EwOKfAkktSda1LQbUkKxSzaOQDHg80BCbYpRyHiR25DiGuyoS4iy3sYLHmPDpY81hrAhx0SSbxFhvgxCHKUe8giE4-i3hYMRs0UTpyWgEmRu19qlxOso-IqjoDqPMZo0xizOp6J1EPIxI8LLjxoZPeQ1jbHRDnoxLxLzVq-PXv46Rq0glAX3qE5oPyJxk0iRTCyVMYnlDiQzLqfJAzBhGhIRwOVIz02GaRQweR5BKBbmADAzdJ7yQAFojTcJYkhDQiHIt2ZwSAYAAYWXQK2YALLoiCpcFGZufAQ4OF2jEZua0ug4FgKKhJd84Y4rLNoRlzLzidM8AAbiAA)

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF


### TESTING INSTRUCTIONS
create a mixed chart on the explore menu using confidence intervals where the names__yhat_lower bound is negative at least once.

Verify that confidence Intervals behave as Intended.

**names__yhat**: trend/point estimate
**names__yhat_lower**: lower confidence level
**names__yhat_upper**: upper confidence level 

### ADDITIONAL INFORMATION
Fixes #30554
